[AIRFLOW-3888] HA for metastore connection

GuzikJakub opened the pull request:
### Jira

https://issues.apache.org/jira/browse/AIRFLOW-3888

### Description

Creating a connection to a metastore with two hosts for high availability (eg connection 1, connection 2) is not possible because the entire value entered is taken. For our needs, it is necessary to go through subsequent hosts and connect to the first working.

This change allows you to check and then connect to a working metastor.

### Tests

Tested on two metastores, of which one was always inactive

Example:
matastore1(offline); metastoer2(online).
In this case, we will connect to metastore2

### Documentation

The connection is created in the webserver. We separate the metastore with a semicolon, eg: metastror1; metastore2; metastore3

Commit:
HA for Metastore

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Creating a connection to a metasotor with two hosts for high avitablity (eg connection 1, connection 2) is not possible because the entire value entered is taken. For our needs, it is necessary to go through subsequent hosts and connect to the first working.

This change allows you to check and then connect to a working metastor.

GuzikJakub committed:
add function to base_hook

GuzikJakub committed:
update webhdfs_hook

GuzikJakub committed:
back to original version

GuzikJakub committed:
back to original version

XD-DENG commented on airflow/hooks/hive_hooks.py:
```diff
@@ -495,37 +497,49 @@ def get_metastore_client(self):
-
-        socket = TSocket.TSocket(ms.host, ms.port)
+        hosts_list = ms.host.split(';') #split by semicolon
+        for host_name in range(len(hosts_list)):
```
I haven't got time to check the implementation/logic in details yet, but the variable `host_name` here is not making sense to me. It's index actually (from `range(len(...))`), but the name itself is misleading.

Alternatively, you can change this line into `for host_name in hosts_list:` AND change all `hosts_list[host_name]` below into `host_name`.

GuzikJakub committed:
Update hive_hooks.py

Thank you. I made a few changes because during the tests I detected several errors.

I have a question, when I do marge to my pull it will be  still possible to land it in the airflow main branch?

GuzikJakub committed:
Merge pull request #1 from mik-laj/ha-metastore

HA for Metastore

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection 

flake8 code repair

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection 

Flake8 repair

gglanzani commented on airflow/hooks/hive_hooks.py:
```diff
@@ -514,20 +520,37 @@ def get_metastore_client(self):
+            self.log.info("Trying to connect to %s:%s", conn.host, conn.port)
+            try:
+                result = host_socket.connect_ex((conn.host, conn.port))
+                host_socket.close()
```
We can rewrite the try like this
```python
if host_socket.connect_ex((conn.host, conn.port)) == 0:
    self.log.info("Connected to %s:%s", conn.host, conn.port)
    host_socket.close()
    return conn
```

and avoid the `break`, the `results` and `valid_conn` variable, etc.

GuzikJakub replied:
Code corrected and thrown. The wider answer will be prepared below.

I also removed variables such as break, results and valid_conn

gglanzani commented on airflow/hooks/hive_hooks.py:
```diff
@@ -514,20 +520,37 @@ def get_metastore_client(self):
+                result = host_socket.connect_ex((conn.host, conn.port))
+                host_socket.close()
+            except Exception:
+                pass
```
Why are we swallowing this one without at least logging?

I understand that one of the two hosts might fail, but if both fail I won't see anything in the logs.

GuzikJakub replied:
Test (but I made some corrections to the code):
[2019-03-04 14:08:51,954] {hive_hooks.py:528} INFO - Trying to connect to Metastore1
[2019-03-04 14:08:54,960] {hive_hooks.py:534} INFO - Could not connect to Metastore1
[2019-03-04 14:08:54,960] {hive_hooks.py:528} INFO - Trying to connect to Metastore2
[2019-03-04 14:08:54,962] {hive_hooks.py:534} INFO - Could not connect to Metastore2
[2019-03-04 14:08:54,970] {models.py:1760} ERROR - Failed to locate the valid server.
Traceback (most recent call last):
  File "/airflow/models.py", line 1659, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/airflow/sensors/base_sensor_operator.py", line 68, in execute
    while not self.poke(context):
  File "/airflow/sensors/hive_partition_sensor.py", line 73, in poke
    metastore_conn_id=self.metastore_conn_id)
  File "/airflow/hooks/hive_hooks.py", line 468, in __init__
    self.metastore = self.get_metastore_client()
  File "/airflow/hooks/hive_hooks.py", line 491, in get_metastore_client
    raise AirflowException("Failed to locate the valid server.")
airflow.exceptions.AirflowException: Failed to locate the valid server.
[2019-03-04 14:08:54,971] {models.py:1789} INFO - All retries failed; marking task as FAILED

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Code behavior improvements

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Add test

[AIRFLOW-3888] HA for metastore connection

test improvement

[AIRFLOW-3888] HA for metastore connection

test improvement

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection 

Add test

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Add test

[AIRFLOW-3888] HA for metastore connection

test improvement

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

test improvement

Commit:
Merge branch 'master' of https://github.com/GuzikJakub/airflow

Commit:
Merge branch 'master' of https://github.com/GuzikJakub/airflow

pull rebase

Fokko commented on airflow/hooks/hive_hooks.py:
```diff
@@ -498,13 +499,20 @@ def get_metastore_client(self):
         from thrift.protocol import TBinaryProtocol
-        ms = self.metastore_conn
+
+        ms = self._find_vaild_server()
```
valid*

GuzikJakub replied:
Amendment introduced

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Improving the typo in the variable name

gglanzani commented on tests/hooks/test_hive_hook.py:
```diff
@@ -282,6 +283,13 @@ def test_get_max_partition_from_valid_part_specs(self):
         self.assertIsInstance(self.hook.get_metastore_client(), HMSClient)
 
+    @mock.patch("airflow.hooks.hive_hooks.HiveMetastoreHook.get_connection",
+                **{'return_value': [Connection(host="localhost", port=9802)]})
```
Why not `return_value=[Connection(…)]` instead of using the `**` syntax?

mik-laj replied:
Sugested syntax does not work properly in all cases supported by mock. I often use only one syntax, so that I do not have to think which is right at the moment.
Reference: 
https://github.com/apache/airflow/blob/49b98ab2b5dc37c69598d485d62dd39372ab6aa9/tests/contrib/hooks/test_gcp_natural_language_hook.py#L57-L68

gglanzani replied:
> Your scientists were so preoccupied with whether or not they could, they didn’t stop to think if they should.

The suggested syntax does not work, indeed, when you have invalid identifiers as keyword arguments. `return_value` is not one of them

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Mock return_value edit

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Flake8 repair

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Flake8 repair

[AIRFLOW-3888] HA for metastore connection

Test repair

GuzikJakub committed:
[AIRFLOW-3888] HA for metastore connection

Test repair

Commit:
rebase master
